### ISSUE TITLE
[feat] no-unnecessary-key should catch constant keys outside of map calls

### ISSUE THREAD

**Issue by bschlenk** [Rel1cx/eslint-react#1377]
### Problem Description

It's a strange situation and I don't know why someone would write code this way, but I've ran into a few hardcoded string keys that are used outside of a map call. These are not eligible for the "Static keys used to re-mount components" exception the rule docs mention, since these keys are _constant_ they would never trigger a re-mount.

It would be cool if this rule flagged constant keys like this since they meet the criteria for being unnecessary.

### Alternative Solutions

I don't see any other rules in this project that would apply, although in my case the `no-duplicate-keys` rule did catch some that just happened to be duplicates but _also_ strings.

### Feature Details

This would just be an extension on top of the existing no-unnecessary-key rule and enabled by default.

### Examples

```tsx
<Menu.Root>
  <Menu.Option key="one">
    One
  </Menu.Option>
  <Menu.Option key="two">
    Two
  </Menu.Option>
</Menu.Root>
```

Both of these keys are unnecessary:

- they are not used in a `map` call
- they are constants so would never trigger a re-mount

### Evaluation Checklist

- [x] I have had problems that this feature would solve
- [x] I could not find a way to solve the problem with existing features or workarounds
- [x] I have thought very hard about potential edge cases and downsides, and they are acceptable
- [x] I think the feature is well-defined and would provide clear value to users

**Comment by Rel1cx** [comment]
In this type of situation, the key might indeed be completely useless or accidentally set by the developer. However,
> they are not used in a `map` call
> they are constants, so would never trigger a re-mount

The assertions above are insufficient to conclude that the keys are useless. I did not spend much time constructing one that is not in a `map` call and also uses a constant string key, but can still correctly reset the subtree when needed:

```tsx
import { useState } from "react";

function App() {
  const [lang, setLang] = useState<"en" | "fr" | "ar">("en");
  return (
    <div>
      <p>Current language: {lang}</p>
      <IResetChildrenOnWritingDirectionChange1 lang={lang}><input /></IResetChildrenOnWritingDirectionChange1>
      <IResetChildrenOnWritingDirectionChange2 lang={lang}><input /></IResetChildrenOnWritingDirectionChange2>
      <button onClick={() => setLang("en")}>English</button>
      <button onClick={() => setLang("fr")}>French</button>
      <button onClick={() => setLang("ar")}>Arabic</button>
    </div>
  );
}

// In this case, the elements with constant keys themselves do not change position in the JSX syntax tree,
// but conditional expressions (or control flows) select which one is present in that position at that time,
// resulting in only one `div` being rendered at a time but with a different key on it if the direction changes,
// so the input element passed as children gets reset when the direction changes
function IResetChildrenOnWritingDirectionChange1({ lang = "en", children }: { lang?: "en" | "fr" | "ar"; children: React.ReactNode }) {
  return (
    <div>
      {lang === "en" && <div key="ltr">{children}</div>}
      {lang === "fr" && <div key="ltr">{children}</div>}
      {lang === "ar" && <div key="rtl">{children}</div>}
    </div>
  );
}

// Another way to achieve the same effect as above
function IResetChildrenOnWritingDirectionChange2({ lang = "en", children }: { lang?: "en" | "fr" | "ar"; children: React.ReactNode }) {
  if (lang === "ar") return <div key="rtl">{children}</div>;
  return <div key="ltr">{children}</div>;
}

```

**Comment by bschlenk** [comment]
I'm not sure the comment above `IResetChildrenOnWritingDirectionChange1` is true. What react sees in this case is actually:

```tsx
return (
  <div>
    <div key="ltr">{children}</div>
    {false}
    {false}
  </div>
)
```

React will remount the div regardless of the key here, since this is going to be based on the slots that are filled in the jsx, not the final rendered html.

**Comment by bschlenk** [comment]
Here's a [codepen example](https://codepen.io/bschlenk/pen/bNeeqVj?editors=1011). Notice that the ref is called with `null` every time you change the select, regardless of the key. If it worked the way you described, then the ref wouldn't get called when switching between "First" and "Second".

**Comment by bschlenk** [comment]
Sorry I didn't read far enough ahead, I think your `IResetChildrenOnWritingDirectionChange2` is a good counterpoint. It seems like this would be possible to detect though, if you're okay with only being able to catch it only when the different keys are within the same function.

**Comment by Rel1cx** [comment]
> React will remount the `div` regardless of the key here,

The `div` is not important here, i mean

> they can still correctly reset (or preserve) the **subtree** (below the `children` of this `div`) when needed

And the `key` here makes this behavior perform exactly the same in these two examples, regardless of whether the `div` itself is re-mounted or not.

I am sorry that the comment above `IResetChildrenOnWritingDirectionChange1` gave a poor explanation for this, but you can test it here: https://codepen.io/undefiend/pen/zxBBwBd?editors=1010.

Type some text into the input boxes in both examples. When you click the button below to toggle the language between the first two langs, **the `<input>` elements in neither example are reset**.

When you switch to the third `lang`, **the `input` elements in both examples are reset, and the text inside them is cleared**. If you type new text and switch from the third `lang` back to either of the first two, both `input` elements will be reset again.

>
> ```tsx
> return (
>   <div>
>     <div key="ltr">{children}</div>
>     {false}
>     {false}
>   </div>
> )
> ```
>

Here, the `false` is also irrelevant for the demonstrated usage of using a `key` to control the preservation and reset of inputs in the subtree. You can directly replace all `div` and `false` in `1` with `<React.Fragment>` to create a `1a`; the actual effect will remain the same:

```tsx
function IResetChildrenOnWritingDirectionChange1a({ lang = "en", children }: { lang?: "en" | "fr" | "ar"; children: React.ReactNode }) {
  return (
    <>
      {lang === "en" ? <React.Fragment key="ltr">{children}</React.Fragment> : <></>}
      {lang === "fr" ? <React.Fragment key="ltr">{children}</React.Fragment> : <></>}
      {lang === "ar" ? <React.Fragment key="rtl">{children}</React.Fragment> : <></>}
    </>
  );
}
```

**Comment by Rel1cx** [comment]
> It seems like this would be possible to detect though

Yes, but I tend to think that rules shouldn't be too restrictive about some potential use cases that haven't been properly explored yet. It might sound counterintuitive coming from a linter maker, but banning these potential uses from the start would stifle community innovation, especially in the React community.

**Comment by Rel1cx** [comment]
Just got back from #1381, it reminded me that for scenarios like `<div {...props} key="1" />` and `<div key="1" {...props} />`, we also cannot report an unnecessary key.

**Comment by Rel1cx** [comment]
Actually, even this should not be reported as unnecessary: 
```tsx
<Menu key="one">
```
Because the `Menu` component might look like this inside:

```tsx
function Menu({ key }: { key: string }) { /* do something with the key */ }
```

~~In React 19 and later with the new JSX transform, `key` and `ref` are just regular props. They can be defined and used by the user~~. Reporting them as unnecessary might cause them to make changes that create problems which are challenging to detect.

**Comment by Rel1cx** [comment]
~~We now have sufficient reasons to not to enable this rule by default too~~.

BTW, for those interested in what happened in `IResetChildrenOnWritingDirectionChange1` and how React chooses whether to reuse or replace the `<input/>` elements in the example during the reconciliation process, I've created a [playground link](https://jser.pro/ddir/rie?reactVersion=19.0.0-rc-fb9a90fa48-20240614&snippetKey=a887o8kme63y9vbva59) on ddir.

**Comment by bschlenk** [comment]
Sorry for commenting on a closed issue. Not sure if this changes anything, but with the new JSX transform specifically, I don't think key is ever passed down to or readable by children. Even if it gets spread in through props: https://codepen.io/bschlenk/pen/XJKNWeq?editors=1011

**Comment by Rel1cx** [comment]
> Sorry for commenting on a closed issue. Not sure if this changes anything, but with the new JSX transform specifically, I don't think key is ever passed down to or readable by children. Even if it gets spread in through props: https://codepen.io/bschlenk/pen/XJKNWeq?editors=1011

You're right! The child component isn't receiving the key prop, I made a mistake on that. Thank you so much for continuing to follow up on this issue even though it's already closed.

**Comment by Rel1cx** [comment]
I think I can update the rule to check for unnecessary keys in the examples you described at the beginning, while also properly handling some of the edge cases mentioned above. Give me some time, but if anyone is willing to submit a PR, I'd be happy to accept it too.

**Comment by silverwind** [comment]
I notice the rule is now flagging `key` inside `Array<ReactNode>` like in [this example](https://reactplayground.vercel.app/#N4IgLgziBcBmCGAbCBTANCAbrK1QEsA7AExQA8A6AK1xHwFsAHAewCcwACAQUcY9lbN6HAOQUA9D0bUIZEQB1CDFuw4AlFPADGnAUNGtNOhUqZtOwDlsPwwKNc2acAvv0HCRNnQFpiQ8VqI+CiEYCaKWsyEEJyRofBEKKwcALwcfloArvQhYBQA5ihgAKKIKDmhAEIAngCSxAAUno5hAJQRUTEcgk6pVjZ2Dk4NcWAJhEnthD15hiRJDYoc6kZ51pp2peW5i4TLyxraeQDKYKz4OgCyzKRoS-uEmYiId3v7hzoU67YoWxVgDSkU2WU1aIAwUhkZBgdDMqg+uncBlW4QmZBUnFICCenAarVSAD4OMB7oYwJlWHsANr3ZYAHmI+EwHAA1ihqil5CAAIxcgncuniRmYAmvfYcBlM1nszkgABMfLlguFRPuAF1FM5FCBnM4gA). Are these keys safe to remove when the array order is static?
